akka-actor/src/main/scala/akka/pattern/AskSupport.scala

```diff
@@ -207,53 +255,106 @@ final class AskableActorSelection(val actorSel: ActorSelection) extends AnyVal {
 
   def ask(message: Any)(implicit timeout: Timeout, sender: ActorRef = Actor.noSender): Future[Any] =
-    internalAsk(message, timeout, sender)
+    internalAsk(_ ⇒ message, timeout, sender)
 
   /**
    * INTERNAL API: for binary compatibility
    */
   protected def ?(message: Any)(implicit timeout: Timeout): Future[Any] =
-    internalAsk(message, timeout, ActorRef.noSender)
+    internalAsk(_ ⇒ message, timeout, ActorRef.noSender)
 
   def ?(message: Any)(implicit timeout: Timeout, sender: ActorRef = Actor.noSender): Future[Any] =
-    internalAsk(message, timeout, sender)
+    internalAsk(_ ⇒ message, timeout, sender)
 
   /**
    * INTERNAL API: for binary compatibility
    */
-  private[pattern] def internalAsk(message: Any, timeout: Timeout, sender: ActorRef): Future[Any] = actorSel.anchor match {
+  private[pattern] def internalAsk(messageFactory: ActorRef ⇒ Any, timeout: Timeout, sender: ActorRef): Future[Any] = actorSel.anchor match {
     case ref: InternalActorRef ⇒
-      if (timeout.duration.length <= 0)
+      if (timeout.duration.length <= 0) {
+        val message = messageFactory(ActorRef.noSender)
         Future.failed[Any](
           new IllegalArgumentException(s"""Timeout length must not be negative, question not sent to [$actorSel]. Sender[$sender] sent the message of type "${message.getClass.getName}"."""))
-      else {
-        val a = PromiseActorRef(ref.provider, timeout, targetName = actorSel, message.getClass.getName, sender)
-        actorSel.tell(message, a)
-        a.result.future
+      } else {
+        val promiseRef = new PromiseActorRef(ref.provider)
+        val message = messageFactory(promiseRef)
+        promiseRef.setTimeout(timeout, actorSel, sender, message)
+        actorSel.tell(message, promiseRef)
+        promiseRef.future
       }
-    case _ ⇒ Future.failed[Any](new IllegalArgumentException(s"""Unsupported recipient ActorRef type, question not sent to [$actorSel]. Sender[$sender] sent the message of type "${message.getClass.getName}"."""))
+    case _ ⇒
+      val message = messageFactory(ActorRef.noSender)
+      Future.failed[Any](new IllegalArgumentException(s"""Unsupported recipient ActorRef type, question not sent to [$actorSel]. Sender[$sender] sent the message of type "${message.getClass.getName}"."""))
   }
 }
 
 /**
+ * INTERNAL API
+ */
+private[akka] object PromiseActorRef {
+  private case object Registering
+  private case object Stopped
+  private final case class StoppedWithPath(path: ActorPath)
+
+  private val ActorStopResult = Failure(new ActorKilledException("Stopped"))
+}
+
+/**
  * Akka private optimized representation of the temporary actor spawned to
  * receive the reply to an "ask" operation.
  *
  * INTERNAL API
  */
-private[akka] final class PromiseActorRef private (val provider: ActorRefProvider, val result: Promise[Any], _mcn: String)
-  extends MinimalActorRef {
+private[akka] final class PromiseActorRef(val provider: ActorRefProvider)
```

Comment:
This should probably be renamed to `PromiseRefImpl` for consistency with Akka Typed code and to be less ambiguous with public `PromiseRef`. However I've found mentions of `PromiseActorRef` in documentation and it is also used in public `ActorPromiseActorRef` java API (although I couldn't figure out what's it purpose or how is it supposed to be used) so I've decided to leave it unchanged for now.